### ISSUE TITLE
Issue about forward_gpu call forward_cpu.

### ISSUE THREAD

**Issue by YafeiWangAlice** [BVLC/caffe#6692]
I am walkthroughing caffe-opencl code.

In Layer.hpp, 
  _virtual void Forward_gpu(const vector<Blob<MItype>*>& bottom,
                           const vector<Blob<MOtype>*>& top) {
    // LOG(WARNING) << "Using CPU code as backup.";
    Forward_cpu(bottom, top);
  }_

I think it means when the layer which didn't implement forward_gpu(), than the layer's forward_cpu() will be called instead.

But in the Forward() function (in layer.hpp) as follows:

_template<typename Dtype, typename MItype, typename MOtype>
inline Dtype Layer<Dtype, MItype, MOtype>::Forward(
    const vector<Blob<MItype>*>& bottom,
    const vector<Blob<MOtype>*>& top) {
  Dtype loss = 0;
  Reshape(bottom, top);
  for (int_tp bottom_id = 0; bottom_id < bottom.size(); ++bottom_id) {
    if (bottom_quants_.size() > 0) {
      // Observe values needed for quantization if necessary
      bottom_quants_[bottom_id % bottom_quants_.size()]
            ->ObserveIn(bottom[bottom_id]->count(), bottom[bottom_id]->data());
      // Run pseudo quantization if necessary
      bottom_quants_[bottom_id % bottom_quants_.size()]
            ->PseudoQuantIn(bottom[bottom_id]->count(),
                            bottom[bottom_id]->data(),
                            bottom[bottom_id]->data());
    }
  }
  switch (Caffe::mode()) {
    case Caffe::CPU:
      Forward_cpu(bottom, top);
      for (int_tp top_id = 0; top_id < top.size(); ++top_id) {
        if (!this->loss(top_id)) {
          continue;
        }
        const int_tp count = top[top_id]->count();
        const MOtype* data = top[top_id]->cpu_data();
        const MOtype* loss_weights = top[top_id]->cpu_diff();
        loss += caffe_dot(count, data, loss_weights);
      }
      break;
    case Caffe::GPU:
      Forward_gpu(bottom, top);
#ifndef CPU_ONLY
      for (int_tp top_id = 0; top_id < top.size(); ++top_id) {
        if (!this->loss(top_id)) {
          continue;
        }
        const int_tp count = top[top_id]->count();
        **vptr<const MOtype> data = top[top_id]->gpu_data();
        vptr<const MOtype> loss_weights = top[top_id]->gpu_diff();**
        MOtype blob_loss = 0;
        device_->template dot<MOtype>(count, data, loss_weights, &blob_loss);
        // TODO: Type conversion may be necessary
        // TODO: Compute on CPU if Forward_gpu uses CPU fallback?
        loss += blob_loss;
      }
#endif  // !CPU_ONLY
    break;
  default:
    LOG(FATAL) << "Unknown caffe mode.";
  }
  for (int_tp top_id = 0; top_id < top.size(); ++top_id) {
    if (top_quants_.size() > 0) {
      top_quants_[top_id % top_quants_.size()]
                       ->ObserveOut(top[top_id]->count(), top[top_id]->data());
      top_quants_[top_id % top_quants_.size()]
            ->PseudoQuantOut(top[top_id]->count(),
                             top[top_id]->data(),
                             top[top_id]->data());
    }
  }
  for (int_tp blob_id = 0; blob_id < blobs_.size(); ++blob_id) {
    if (blobs_quants_.size() > 0) {
      blobs_quants_[blob_id % blobs_quants_.size()]
               ->ObserveOut(blobs_[blob_id]->count(), blobs_[blob_id]->data());
    }
  }
  return loss;
}_

After  Forward_gpu(bottom, top) called, only gpu_data() is handled no cpu_data() is there.
What if in forward_gpu() is actually called forward_cpu()? or is it the TODO thing?
